### ISSUE TITLE
nan Gradient in kornia.geometry.rotation_matrix_to_quaternion

### ISSUE THREAD

**Issue by simon-schaefer** [kornia/kornia#1653]
### Describe the bug

When using `torch.float32` the transformation from rotation matrices to quaternions outputs nan valued gradients. This is caused by `trace + 1.0` (in `sq = ...`) which might be slightly below zero, for single precision. It should be `sq = torch.sqrt(trace + 1.0 + eps) * 2.0` instead. 

```
    def trace_positive_cond():
        sq = torch.sqrt(trace + 1.0) * 2.0  # sq = 4 * qw.
        qw = 0.25 * sq
        qx = safe_zero_division(m21 - m12, sq)
        qy = safe_zero_division(m02 - m20, sq)
        qz = safe_zero_division(m10 - m01, sq)
        if order == QuaternionCoeffOrder.XYZW:
            return torch.cat((qx, qy, qz, qw), dim=-1)
        return torch.cat((qw, qx, qy, qz), dim=-1)
```

### Reproduction steps

```bash
import kornia
import torch


def main():
    R = torch.tensor([[1, 0, 0],
                      [0, -1, 0],
                      [0, 0, -1]], dtype=torch.float32, requires_grad=True)
    q = kornia.geometry.rotation_matrix_to_quaternion(R)
    print(torch.autograd.grad(q.sum(), R))


if __name__ == '__main__':
    main()
```

### Expected behavior

The code above outputs a gradient including nan values. As the rotation matrix to quaternion transform includes safe divisions, it should not yield nan values.

```
(tensor([[    nan,  0.2500,  0.2500],
        [ 0.2500,     nan, -0.2500],
        [ 0.2500,  0.2500,     nan]]),)

```

### Environment

```shell
python 3.8.10
pytorch 1.11.0
kornia 0.6.2
```

**Comment by Akaame** [comment]
```python
quaternion: torch.Tensor = torch.where(trace > 0.0, trace_positive_cond(), where_1)
return quaternion
```
`trace_positive_cond` path is not taken for this matrix. What do you get when you calculate `cond_1` by hand?

**Comment by edgarriba** [comment]
@simon-schaefer can you guys send a quick PR adding the `eps` parameter ? Besides, I'm recently exploring alternatives to handle geometry conversions via Lie Algebra. I have an internal prototype of a python (not only pytorch) implementation of hauke's library Sophus https://github.com/strasdat/Sophus that I hope we can release at some point soon.

**Comment by simon-schaefer** [comment]
Thank you for your quick reponse!

@Akaame: Analytically, the gradient of `q.sum()` w.r.t. `m00` is 
```math
\fraq{0.25⋅(m₀₀ - m₀₁ - m₀₂ - m₁₀ - m₁₁ + m₁₂ - m₂₀ - m₂₁ - m₂₂ + 1.0)}{(m₀₀ - m₁₁ - m₂₂ + 1.0)^{3/2}}
```
which results in 0.125. It is similar for the other terms, just with a different sign. Thus, the correct outcome is
```
(tensor([[ 0.1250,  0.2500,  0.2500],
        [ 0.2500, -0.1250, -0.2500],
        [ 0.2500,  0.2500, -0.1250]]),)
```
We get the same result when we perturb the initial rotation matrix from above by adding a small eps to the diagonal elements `m11` and `m22`. 

@edgarriba: Sure, will do it. Are you looking for something like https://github.com/princeton-vl/lietorch?

**Comment by simon-schaefer** [comment]
I have submitted the PR for adding the eps.  
https://github.com/kornia/kornia/pull/1665

**Comment by ducha-aiki** [comment]
Solved in #1665
